File: sdk/node/rollout.ts
/** Framework-aware fused rollout client for local training loops. */

import { createHash } from 'node:crypto';
import {
  closeSync,
  lstatSync,
  openSync,
  readFileSync,
  readSync,
  readdirSync,
  realpathSync,
  statSync,
} from 'node:fs';
import { isAbsolute, relative, resolve, sep } from 'node:path';

export class RolloutError extends Error {
  constructor(
    readonly status: number,
    readonly code: string,
    message: string,
  ) {
    super(`${code}: ${message}`);
    this.name = 'RolloutError';
  }
}

export interface RolloutPolicyInfo {
  policy: string;
  version: string;
  adapterSha256: string;
  backendModel: string;
  source: 'filesystem' | 'device';
  current: boolean;
  activeRequests: number;
  retiring: boolean;
}

export interface RolloutExecutorInfo {
  name: string;
  backend: 'vllm';
  endpoint: string;
  adapterRoot: string;
  deviceAdapterSocket?: string;
  fallbackPool?: string;
  maxConcurrentRequests: number;
  maxQueueDepth: number;
  requestTimeoutSecs: number;
  activeRequests: number;
  queuedRequests: number;
  accepting: boolean;
  policies: RolloutPolicyInfo[];
  capabilities: string[];
}

export interface EnsureRolloutExecutorOptions {
  endpoint: string;
  adapterRoot: string;
  deviceAdapterSocket?: string;
  fallbackPool?: string;
  maxConcurrentRequests?: number;
  maxQueueDepth?: number;
  requestTimeoutSecs?: number;
}

export interface RolloutSampling {
  n?: number;
  maxTokens: number;
  temperature?: number;
  topP?: number;
  topK?: number;
  minP?: number;
  repetitionPenalty?: number;
  seed?: number;
  logprobs?: number;
  promptLogprobs?: number;
}

export type RolloutPrompt = string | readonly number[];

export interface RolloutCohort {
  id: string;
  size: number;
  maxWaitMs?: number;
}

export interface RolloutJob {
  idempotencyKey: string;
  policy: string;
  version?: string;
  prompts: Array<{ text: string } | { tokenIds: number[] }>;
  sampling: RolloutSampling;
  deadlineMs?: number;
  cohort?: RolloutCohort;
}

export interface RolloutCompletion {
  index: number;
  text: string;
  tokenIds?: number[];
  promptTokenIds?: number[];
  logprobs?: unknown;
  finishReason?: string;
  stopReason?: unknown;
}

export interface RolloutGenerateResponse {
  executor: string;
  policy: string;
  version: string;
  backendRequestId: string;
  choices: RolloutCompletion[];
  usage: { promptTokens: number; completionTokens: number; totalTokens: number };
  cached: boolean;
}

export interface RolloutBatchItem {
  idempotencyKey: string;
  response?: RolloutGenerateResponse;
  errorCode?: string;
  error?: string;
}

export interface BuildRolloutJobOptions {
  idempotencyKey: string;
  policy: string;
  version?: string;
  prompts: readonly RolloutPrompt[];
  sampling: RolloutSampling;
  deadlineMs?: number;
  cohort?: RolloutCohort;
}

export interface RolloutClientOptions {
  timeoutMs?: number;
  headers?: Record<string, string>;
  fetch?: typeof fetch;
  bearerToken?: string;
  forkEnvPath?: string;
  autoForkCohort?: boolean;
  autoForkCohortMaxWaitMs?: number;
}

const FORK_ENV_PATH = '/etc/smolvm/fork-env';
const ROLLOUT_URL_ENV = 'SMOLVM_ROLLOUT_URL';
const ROLLOUT_TOKEN_ENV = 'SMOLVM_ROLLOUT_TOKEN';
const ROLLOUT_EXECUTOR_ENV = 'SMOLVM_ROLLOUT_EXECUTOR';
const ROLLOUT_POLICY_ENV = 'SMOLVM_ROLLOUT_POLICY';
const ROLLOUT_LEASE_KEYS = [
  ROLLOUT_URL_ENV,
  ROLLOUT_TOKEN_ENV,
  ROLLOUT_EXECUTOR_ENV,
  ROLLOUT_POLICY_ENV,
] as const;

function readForkEnv(path: string): Record<string, string> {
  let contents: string;
  try {
    contents = readFileSync(path, 'utf8');
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') return {};
    throw new Error(`cannot read smolvm fork assignment ${path}: ${String(error)}`);
  }
  const values: Record<string, string> = {};
  contents.split(/\r?\n/).forEach((line, index) => {
    if (line.length === 0 || line.startsWith('#')) return;
    const separator = line.indexOf('=');
    if (separator <= 0) {
      throw new Error(`invalid smolvm fork assignment at ${path}:${index + 1}`);
    }
    const key = line.slice(0, separator);
    if (Object.prototype.hasOwnProperty.call(values, key)) {
      throw new Error(`duplicate ${key} in smolvm fork assignment ${path}`);
    }
    values[key] = line.slice(separator + 1);
  });
  return values;
}

function leaseConfiguration(path: string) {
  const values = readForkEnv(path);
  for (const key of ROLLOUT_LEASE_KEYS) {
    if (process.env[key] !== undefined) values[key] = process.env[key] as string;
  }
  const missing = ROLLOUT_LEASE_KEYS.filter((key) => !values[key]);
  if (missing.length > 0) {
    throw new Error(`smolvm rollout lease configuration is unavailable: missing ${missing.join(', ')}`);
  }
  const url = values[ROLLOUT_URL_ENV].replace(/\/+$/, '');
  const executor = values[ROLLOUT_EXECUTOR_ENV];
  const suffix = `/rollout-executors/${executor}`;
  if (!url.endsWith(suffix)) {
    throw new Error(`${ROLLOUT_URL_ENV} does not match ${ROLLOUT_EXECUTOR_ENV}`);
  }
  const apiUrl = url.slice(0, -suffix.length);
  if (!apiUrl.startsWith('http://') && !apiUrl.startsWith('https://')) {
    throw new Error(`${ROLLOUT_URL_ENV} must be an HTTP URL`);
  }
  return {
    apiUrl,
    executor,
    bearerToken: values[ROLLOUT_TOKEN_ENV],
    leasePolicy: values[ROLLOUT_POLICY_ENV],
    assignment: values,
  };
}

function lengthBytes(value: number): Buffer {
  const output = Buffer.allocUnsafe(8);
  output.writeBigUInt64LE(BigInt(value));
  return output;
}

/** Deterministically hash adapter file names, sizes, and contents. */
export function adapterSha256(directory: string): string {
  const root = realpathSync(directory);
  if (!statSync(root).isDirectory()) throw new Error('adapter path must be a directory');
  const files: string[] = [];
  const walk = (current: string) => {
    for (const entry of readdirSync(current, { withFileTypes: true })) {
      const path = resolve(current, entry.name);
      if (entry.isSymbolicLink() || lstatSync(path).isSymbolicLink()) {
        throw new Error('adapter directories cannot contain symlinks');
      }
      if (entry.isDirectory()) walk(path);
      else if (entry.isFile()) files.push(path);
      else throw new Error('adapter directories may contain only regular files');
      if (files.length > 4096) throw new Error('adapter contains more than 4096 files');
    }
  };
  walk(root);
  if (files.length === 0) throw new Error('adapter directory contains no files');
  files.sort((left, right) =>
    Buffer.compare(
      Buffer.from(relative(root, left).split(sep).join('/')),
      Buffer.from(relative(root, right).split(sep).join('/')),
    ),
  );

  const hash = createHash('sha256');
  const buffer = Buffer.allocUnsafe(1024 * 1024);
  let total = 0;
  for (const path of files) {
    const name = Buffer.from(relative(root, path).split(sep).join('/'));
    const size = statSync(path).size;
    total += size;
    if (total > 32 * 1024 * 1024 * 1024) throw new Error('adapter exceeds 32 GiB');
    hash.update(lengthBytes(name.length));
    hash.update(name);
    hash.update(lengthBytes(size));
    const fd = openSync(path, 'r');
    try {
      let offset = 0;
      while (offset < size) {
        const count = readSync(fd, buffer, 0, Math.min(buffer.length, size - offset), offset);
        if (count === 0) throw new Error(`adapter file changed while hashing: ${path}`);
        hash.update(buffer.subarray(0, count));
        offset += count;
      }
    } finally {
      closeSync(fd);
    }
  }
  return hash.digest('hex');
}

function segment(value: string): string {
  return encodeURIComponent(value);
}

function optional<T extends object>(value: T): T {
  return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined)) as T;
}

/** Client for publishing LoRA versions and generating fused policy cohorts. */
export class RolloutClient {
  readonly apiUrl: string;
  readonly executor: string;
  readonly leasePolicy: string | undefined;
  private readonly timeoutMs: number;
  private readonly headers: Record<string, string>;
  private readonly fetchFn: typeof fetch;
  private readonly autoForkCohort: boolean;
  private readonly autoForkCohortMaxWaitMs?: number;
  private readonly forkAssignment: Record<string, string>;
  private forkCohortRound = 0;
  private readonly forkCohorts = new Map<string, RolloutCohort>();

  constructor(apiUrl?: string, executor?: string, options: RolloutClientOptions = {}) {
    let bearerToken = options.bearerToken;
    let forkAssignment: Record<string, string> = {};
    let leasePolicy: string | undefined;
    if (apiUrl === undefined && executor === undefined) {
      const discovered = leaseConfiguration(options.forkEnvPath ?? FORK_ENV_PATH);
      apiUrl = discovered.apiUrl;
      executor = discovered.executor;
      leasePolicy = discovered.leasePolicy;
      forkAssignment = discovered.assignment;
      if (bearerToken !== undefined && bearerToken !== discovered.bearerToken) {
        throw new Error('bearerToken conflicts with the smolvm lease credential');
      }
      bearerToken = discovered.bearerToken;
    } else if (apiUrl === undefined || executor === undefined) {
      throw new TypeError('apiUrl and executor must be provided together');
    }
    const autoForkCohort = options.autoForkCohort ?? true;
    if (typeof autoForkCohort !== 'boolean') {
      throw new TypeError('autoForkCohort must be a boolean');
    }
    const maxWaitMs = options.autoForkCohortMaxWaitMs ?? 250;
    if (!Number.isInteger(maxWaitMs) || maxWaitMs < 1 || maxWaitMs > 60_000) {
      throw new RangeError('autoForkCohortMaxWaitMs must be between 1 and 60000');
    }
    this.apiUrl = apiUrl.replace(/\/+$/, '');
    this.executor = executor;
    this.leasePolicy = leasePolicy;
    this.timeoutMs = options.timeoutMs ?? 300_000;
    this.headers = {
      ...(options.headers ?? {}),
      ...(bearerToken === undefined ? {} : { authorization: `Bearer ${bearerToken}` }),
    };
    this.fetchFn = options.fetch ?? fetch;
    this.autoForkCohort = autoForkCohort;
    this.autoForkCohortMaxWaitMs = maxWaitMs;
    this.forkAssignment = forkAssignment;
  }

  private automaticForkCohort(idempotencyKey: string): RolloutCohort | undefined {
    if (!this.autoForkCohort) return undefined;
    const batchId = process.env.SMOLVM_FORK_BATCH_ID ?? this.forkAssignment.SMOLVM_FORK_BATCH_ID;
    const batchSizeText = process.env.SMOLVM_FORK_BATCH_SIZE
      ?? this.forkAssignment.SMOLVM_FORK_BATCH_SIZE;
    if (batchId === undefined && batchSizeText === undefined) return undefined;
    if (!batchId || batchSizeText === undefined) {
      throw new Error('SMOLVM_FORK_BATCH_ID and SMOLVM_FORK_BATCH_SIZE must be set together');
    }
    const batchSize = Number(batchSizeText);
    if (!Number.isInteger(batchSize) || batchSize < 2 || batchSize > 1024) {
      throw new RangeError('SMOLVM_FORK_BATCH_SIZE must be between 2 and 1024');
    }
    let groupIndex = 0;
    let groupSize = batchSize;
    if (batchSize > 256) {
      const forkIndexText = process.env.SMOLVM_FORK_INDEX ?? this.forkAssignment.SMOLVM_FORK_INDEX;
      const forkIndex = Number(forkIndexText);
      if (
        forkIndexText === undefined
        || !Number.isInteger(forkIndex)
        || forkIndex < 0
        || forkIndex >= batchSize
      ) {
        throw new RangeError('SMOLVM_FORK_INDEX is required and must be inside the fork batch');
      }
      groupIndex = Math.floor(forkIndex / 256);
      groupSize = Math.min(256, batchSize - groupIndex * 256);
    }
    const cached = this.forkCohorts.get(idempotencyKey);
    if (cached !== undefined) {
      this.forkCohorts.delete(idempotencyKey);
      this.forkCohorts.set(idempotencyKey, cached);
      return cached;
    }
    const round = this.forkCohortRound++;
    const digest = createHash('sha256')
      .update(`${batchId}\0${this.executor}\0${groupIndex}\0${round}`)
      .digest('hex');
    const cohort = optional({
      id: `fork-${digest.slice(0, 32)}`,
      size: groupSize,
      maxWaitMs: this.autoForkCohortMaxWaitMs,
    }) as RolloutCohort;
    this.forkCohorts.set(idempotencyKey, cohort);
    if (this.forkCohorts.size > 1024) {
      this.forkCohorts.delete(this.forkCohorts.keys().next().value as string);
    }
    return cohort;
  }

  private get executorPath(): string {
    return `/rollout-executors/${segment(this.executor)}`;
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchFn(`${this.apiUrl}${path}`, {
        method,
        headers: { ...this.headers, 'content-type': 'application/json' },
        ...(body === undefined ? {} : { body: JSON.stringify(body) }),
        signal: controller.signal,
      });
      const text = await response.text();
      let payload: unknown;
      try {
        payload = text.length === 0 ? undefined : JSON.parse(text);
      } catch {
        payload = undefined;
      }
      if (!response.ok) {
        const error = payload as { code?: unknown; error?: unknown } | undefined;
        throw new RolloutError(
          response.status,
          typeof error?.code === 'string' ? error.code : 'HTTP_ERROR',
          typeof error?.error === 'string' ? error.error : text || response.statusText,
        );
      }
      return payload as T;
    } catch (error) {
      if (error instanceof RolloutError) throw error;
      const message = error instanceof Error ? error.message : String(error);
      throw new RolloutError(0, 'UNAVAILABLE', message);
    } finally {
      clearTimeout(timer);
    }
  }

  async ensureVllmExecutor(options: EnsureRolloutExecutorOptions): Promise<RolloutExecutorInfo> {
    const desired = {
      name: this.executor,
      backend: 'vllm',
      endpoint: options.endpoint,
      adapterRoot: realpathSync(options.adapterRoot),
      maxConcurrentRequests: options.maxConcurrentRequests ?? 32,
      maxQueueDepth: options.maxQueueDepth ?? 256,
      requestTimeoutSecs: options.requestTimeoutSecs ?? 300,
      ...(options.deviceAdapterSocket === undefined
        ? {}
        : { deviceAdapterSocket: realpathSync(options.deviceAdapterSocket) }),
      ...(options.fallbackPool === undefined ? {} : { fallbackPool: options.fallbackPool }),
    };
    try {
      return await this.request<RolloutExecutorInfo>('POST', '/rollout-executors', desired);
    } catch (error) {
      if (!(error instanceof RolloutError) || error.status !== 409) throw error;
    }
    const current = await this.info();
    const comparable = {
      backend: current.backend,
      endpoint: current.endpoint,
      adapterRoot: current.adapterRoot,
      deviceAdapterSocket: current.deviceAdapterSocket,
      fallbackPool: current.fallbackPool,
      maxConcurrentRequests: current.maxConcurrentRequests,
      maxQueueDepth: current.maxQueueDepth,
      requestTimeoutSecs: current.requestTimeoutSecs,
    };
    const expected = {
      backend: 'vllm',
      endpoint: desired.endpoint,
      adapterRoot: desired.adapterRoot,
      deviceAdapterSocket: options.deviceAdapterSocket === undefined
        ? undefined
        : desired.deviceAdapterSocket,
      fallbackPool: options.fallbackPool,
      maxConcurrentRequests: desired.maxConcurrentRequests,
      maxQueueDepth: desired.maxQueueDepth,
      requestTimeoutSecs: desired.requestTimeoutSecs,
    };
    if (JSON.stringify(comparable) !== JSON.stringify(expected)) {
      throw new RolloutError(
        409,
        'CONFLICT',
        `executor ${JSON.stringify(this.executor)} exists with different configuration`,
      );
    }
    return current;
  }

  info(): Promise<RolloutExecutorInfo> {
    return this.request('GET', this.executorPath);
  }

  async publishPolicy(
    policy: string,
    version: string,
    adapterDirectory: string,
    retainPrevious = false,
  ): Promise<RolloutPolicyInfo> {
    const info = await this.info();
    const root = realpathSync(info.adapterRoot);
    const adapter = realpathSync(adapterDirectory);
    const adapterPath = relative(root, adapter);
    if (
      adapterPath.length === 0 ||
      adapterPath === '..' ||
      adapterPath.startsWith(`..${sep}`) ||
      isAbsolute(adapterPath)
    ) {
      throw new Error('adapter must be beneath the executor adapter root');
    }
    return this.request('POST', `${this.executorPath}/policies`, {
      policy,
      version,
      adapterPath: adapterPath.split(sep).join('/'),
      adapterSha256: adapterSha256(adapter),
      retainPrevious,
    });
  }

  async publishDevicePolicy(
    policy: string,
    version: string,
    tensorBundleToken: string | Uint8Array,
    retainPrevious = false,
  ): Promise<RolloutPolicyInfo> {
    let token: string;
    if (typeof tensorBundleToken === 'string') {
      if (!/^[0-9a-fA-F]{64}$/.test(tensorBundleToken)) {
        throw new Error('tensor bundle token must be exactly 32 hexadecimal bytes');
      }
      token = tensorBundleToken.toLowerCase();
    } else {
      if (tensorBundleToken.byteLength !== 32) {
        throw new Error('tensor bundle token must contain exactly 32 bytes');
      }
      token = Buffer.from(tensorBundleToken).toString('hex');
    }
    const path = `${this.executorPath}/device-policies`;
    const body = { policy, version, tensorBundleToken: token, retainPrevious };
    try {
      return await this.request('POST', path, body);
    } catch (error) {
      if (!(error instanceof RolloutError) || error.status !== 0) throw error;
      // The controller remembers the consumed token's digest, making an
      // ambiguous transport retry idempotent for this exact publication.
      return this.request('POST', path, body);
    }
  }

  async retirePolicy(policy: string, version: string): Promise<void> {
    await this.request(
      'DELETE',
      `${this.executorPath}/policies/${segment(policy)}/${segment(version)}`,
    );
  }

  job(options: BuildRolloutJobOptions): RolloutJob {
    if (options.prompts.length === 0) throw new Error('at least one prompt is required');
    const text = typeof options.prompts[0] === 'string';
    if (!options.prompts.every((prompt) => (typeof prompt === 'string') === text)) {
      throw new Error('one request cannot mix text and token prompts');
    }
    if (options.cohort !== undefined) {
      if (!options.cohort.id) throw new Error('cohort id cannot be empty');
      if (!Number.isInteger(options.cohort.size) || options.cohort.size < 1 || options.cohort.size > 256) {
        throw new RangeError('cohort size must be between 1 and 256');
      }
      if (
        options.cohort.maxWaitMs !== undefined
        && (!Number.isInteger(options.cohort.maxWaitMs)
          || options.cohort.maxWaitMs < 1
          || options.cohort.maxWaitMs > 60_000)
      ) {
        throw new RangeError('cohort maxWaitMs must be between 1 and 60000');
      }
    }
    return optional({
      idempotencyKey: options.idempotencyKey,
      policy: options.policy,
      version: options.version,
      prompts: options.prompts.map((prompt) =>
        typeof prompt === 'string' ? { text: prompt } : { tokenIds: [...prompt] },
      ),
      sampling: optional({ ...options.sampling, n: options.sampling.n ?? 1 }),
      deadlineMs: options.deadlineMs,
      cohort: options.cohort,
    }) as RolloutJob;
  }

  generate(options: BuildRolloutJobOptions): Promise<RolloutGenerateResponse> {
    const cohort = options.cohort ?? this.automaticForkCohort(options.idempotencyKey);
    return this.request(
      'POST',
      `${this.executorPath}/generate`,
      this.job({ ...options, ...(cohort === undefined ? {} : { cohort }) }),
    );
  }

  async generateBatch(jobs: readonly RolloutJob[]): Promise<RolloutBatchItem[]> {
    const response = await this.request<{ jobs: RolloutBatchItem[] }>(
      'POST',
      `${this.executorPath}/batches`,
      { jobs },
    );
    return response.jobs;
  }

  async close(): Promise<void> {
    await this.request('DELETE', this.executorPath);
  }
}

File: sdk/python/python/smol/rollout.py
"""Framework-aware fused rollout client for TRL, Unsloth, and custom RL loops."""

from __future__ import annotations

import hashlib
import json
import os
import ssl
import struct
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Any, Iterable, Optional, Sequence, Union


class RolloutError(RuntimeError):
    """A structured error returned by the rollout service."""

    def __init__(self, status: int, code: str, message: str):
        super().__init__(f"{code}: {message}")
        self.status = status
        self.code = code
        self.message = message


def adapter_sha256(directory: Union[str, os.PathLike]) -> str:
    """Return the deterministic directory digest required for publication."""

    root = Path(directory).resolve(strict=True)
    if not root.is_dir():
        raise ValueError("adapter path must be a directory")
    files = []
    for current, directories, names in os.walk(root, followlinks=False):
        current_path = Path(current)
        for name in directories:
            if (current_path / name).is_symlink():
                raise ValueError("adapter directories cannot contain symlinks")
        for name in names:
            path = current_path / name
            if path.is_symlink() or not path.is_file():
                raise ValueError("adapter directories may contain only regular files")
            files.append(path)
            if len(files) > 4096:
                raise ValueError("adapter contains more than 4096 files")
    if not files:
        raise ValueError("adapter directory contains no files")

    total = 0
    digest = hashlib.sha256()
    for path in sorted(files, key=lambda item: item.relative_to(root).as_posix()):
        name = path.relative_to(root).as_posix().encode("utf-8")
        size = path.stat().st_size
        total += size
        if total > 32 * 1024 * 1024 * 1024:
            raise ValueError("adapter exceeds 32 GiB")
        digest.update(struct.pack("<Q", len(name)))
        digest.update(name)
        digest.update(struct.pack("<Q", size))
        with path.open("rb") as stream:
            while True:
                chunk = stream.read(1024 * 1024)
                if not chunk:
                    break
                digest.update(chunk)
    return digest.hexdigest()


def _segment(value: str) -> str:
    return urllib.parse.quote(value, safe="")


class RolloutClient:
    """Synchronous client for the framework generation boundary."""

    def __init__(
        self,
        api_url: str,
        executor: str,
        *,
        timeout: float = 300.0,
        ssl_context: Optional[ssl.SSLContext] = None,
    ) -> None:
        self.api_url = api_url.rstrip("/")
        self.executor = executor
        self.timeout = timeout
        self.ssl_context = ssl_context

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[dict[str, Any]] = None,
    ) -> Any:
        data = None if body is None else json.dumps(body, separators=(",", ":")).encode()
        request = urllib.request.Request(
            f"{self.api_url}{path}",
            data=data,
            method=method,
            headers={"content-type": "application/json"},
        )
        try:
            with urllib.request.urlopen(
                request, timeout=self.timeout, context=self.ssl_context
            ) as response:
                payload = response.read()
                return None if not payload else json.loads(payload)
        except urllib.error.HTTPError as error:
            raw = error.read()
            try:
                payload = json.loads(raw)
            except (UnicodeDecodeError, json.JSONDecodeError):
                payload = {"code": "HTTP_ERROR", "error": raw.decode(errors="replace")}
            raise RolloutError(
                error.code,
                str(payload.get("code", "HTTP_ERROR")),
                str(payload.get("error", error.reason)),
            ) from error
        except urllib.error.URLError as error:
            raise RolloutError(0, "UNAVAILABLE", str(error.reason)) from error

    @property
    def _executor_path(self) -> str:
        return f"/rollout-executors/{_segment(self.executor)}"

    def ensure_vllm_executor(
        self,
        *,
        endpoint: str,
        adapter_root: Union[str, os.PathLike],
        device_adapter_socket: Optional[Union[str, os.PathLike]] = None,
        fallback_pool: Optional[str] = None,
        max_concurrent_requests: int = 32,
        max_queue_depth: int = 256,
        request_timeout_secs: int = 300,
    ) -> dict[str, Any]:
        """Create the executor or verify an identical registration."""

        desired: dict[str, Any] = {
            "name": self.executor,
            "backend": "vllm",
            "endpoint": endpoint,
            "adapterRoot": str(Path(adapter_root).resolve()),
            "maxConcurrentRequests": max_concurrent_requests,
            "maxQueueDepth": max_queue_depth,
            "requestTimeoutSecs": request_timeout_secs,
        }
        if fallback_pool is not None:
            desired["fallbackPool"] = fallback_pool
        if device_adapter_socket is not None:
            desired["deviceAdapterSocket"] = str(Path(device_adapter_socket).resolve())
        try:
            return self._request("POST", "/rollout-executors", desired)
        except RolloutError as error:
            if error.status != 409:
                raise
        current = self.info()
        comparable = {
            "backend": current["backend"],
            "endpoint": current["endpoint"],
            "adapterRoot": current["adapterRoot"],
            "deviceAdapterSocket": current.get("deviceAdapterSocket"),
            "fallbackPool": current.get("fallbackPool"),
            "maxConcurrentRequests": current["maxConcurrentRequests"],
            "maxQueueDepth": current["maxQueueDepth"],
            "requestTimeoutSecs": current["requestTimeoutSecs"],
        }
        expected = {
            "backend": "vllm",
            "endpoint": endpoint,
            "adapterRoot": desired["adapterRoot"],
            "deviceAdapterSocket": desired.get("deviceAdapterSocket"),
            "fallbackPool": fallback_pool,
            "maxConcurrentRequests": max_concurrent_requests,
            "maxQueueDepth": max_queue_depth,
            "requestTimeoutSecs": request_timeout_secs,
        }
        if comparable != expected:
            raise RolloutError(
                409,
                "CONFLICT",
                f"executor {self.executor!r} exists with different configuration",
            )
        return current

    def info(self) -> dict[str, Any]:
        """Return capabilities, queue state, fallback, and published policies."""

        return self._request("GET", self._executor_path)

    def publish_policy(
        self,
        policy: str,
        version: str,
        adapter_directory: Union[str, os.PathLike],
        *,
        retain_previous: bool = False,
    ) -> dict[str, Any]:
        """Content-verify and atomically publish one immutable LoRA version."""

        info = self.info()
        root = Path(info["adapterRoot"]).resolve(strict=True)
        adapter = Path(adapter_directory).resolve(strict=True)
        try:
            relative = adapter.relative_to(root)
        except ValueError as error:
            raise ValueError("adapter must be beneath the executor adapter root") from error
        return self._request(
            "POST",
            f"{self._executor_path}/policies",
            {
                "policy": policy,
                "version": version,
                "adapterPath": relative.as_posix(),
                "adapterSha256": adapter_sha256(adapter),
                "retainPrevious": retain_previous,
            },
        )

    def publish_device_policy(
        self,
        policy: str,
        version: str,
        tensor_bundle_token: Union[bytes, str],
        *,
        retain_previous: bool = False,
    ) -> dict[str, Any]:
        """Atomically publish a one-use device-resident LoRA allocation."""

        token = (
            tensor_bundle_token.hex()
            if isinstance(tensor_bundle_token, bytes)
            else tensor_bundle_token
        )
        if not isinstance(token, str) or len(token) != 64:
            raise ValueError("tensor bundle token must contain exactly 32 bytes")
        try:
            token = bytes.fromhex(token).hex()
        except ValueError as error:
            raise ValueError("tensor bundle token must be hexadecimal") from error
        path = f"{self._executor_path}/device-policies"
        body = {
            "policy": policy,
            "version": version,
            "tensorBundleToken": token,
            "retainPrevious": retain_previous,
        }
        try:
            return self._request("POST", path, body)
        except RolloutError as error:
            if error.status != 0:
                raise
            # The controller remembers the consumed token's digest, making an
            # ambiguous transport retry idempotent for this exact publication.
            return self._request("POST", path, body)

    def retire_policy(self, policy: str, version: str) -> None:
        """Stop routing and unload a policy version after requests drain."""

        self._request(
            "DELETE",
            f"{self._executor_path}/policies/{_segment(policy)}/{_segment(version)}",
        )

    @staticmethod
    def _prompts(
        prompts: Union[Sequence[str], Sequence[Sequence[int]]]
    ) -> list[dict[str, Any]]:
        if not prompts:
            raise ValueError("at least one prompt is required")
        if isinstance(prompts[0], str):
            if not all(isinstance(prompt, str) for prompt in prompts):
                raise ValueError("one request cannot mix text and token prompts")
            return [{"text": prompt} for prompt in prompts]
        if not all(not isinstance(prompt, str) for prompt in prompts):
            raise ValueError("one request cannot mix text and token prompts")
        return [{"tokenIds": list(prompt)} for prompt in prompts]

    def job(
        self,
        *,
        idempotency_key: str,
        policy: str,
        prompts: Union[Sequence[str], Sequence[Sequence[int]]],
        max_tokens: int,
        version: Optional[str] = None,
        n: int = 1,
        deadline_ms: Optional[int] = None,
        **sampling: Any,
    ) -> dict[str, Any]:
        """Build a generation job for ``generate`` or a fused cohort."""

        parameters = {"n": n, "maxTokens": max_tokens}
        wire_names = {
            "temperature": "temperature",
            "top_p": "topP",
            "top_k": "topK",
            "min_p": "minP",
            "repetition_penalty": "repetitionPenalty",
            "seed": "seed",
            "logprobs": "logprobs",
            "prompt_logprobs": "promptLogprobs",
        }
        unknown = set(sampling) - set(wire_names)
        if unknown:
            raise TypeError(f"unknown sampling parameters: {sorted(unknown)}")
        parameters.update(
            {wire_names[name]: value for name, value in sampling.items() if value is not None}
        )
        job: dict[str, Any] = {
            "idempotencyKey": idempotency_key,
            "policy": policy,
            "prompts": self._prompts(prompts),
            "sampling": parameters,
        }
        if version is not None:
            job["version"] = version
        if deadline_ms is not None:
            job["deadlineMs"] = deadline_ms
        return job

    def generate(self, **job: Any) -> dict[str, Any]:
        """Generate through one policy, returning exact token IDs and logprobs."""

        return self._request("POST", f"{self._executor_path}/generate", self.job(**job))

    def generate_batch(self, jobs: Iterable[dict[str, Any]]) -> list[dict[str, Any]]:
        """Submit independent policy jobs together for backend fusion."""

        response = self._request(
            "POST", f"{self._executor_path}/batches", {"jobs": list(jobs)}
        )
        return response["jobs"]

    def close(self) -> None:
        """Drain and delete the executor registration."""

        self._request("DELETE", self._executor_path)
